packages/components/src/components/SquiggleViewer/SqViewNode.tsx

```diff
@@ -0,0 +1,166 @@
+import { SqValue, SqValuePath } from "@quri/squiggle-lang";
+
+import { getChildrenValues, TraverseCalculatorEdge } from "./utils.js";
+
+//We might want to bring this into the SquiggleLang library. The ``traverseCalculatorEdge`` part is awkward though.
+class SqValueNode {
+  constructor(
+    public root: SqValue,
+    public path: SqValuePath,
+    public traverseCalculatorEdge: TraverseCalculatorEdge
+  ) {}
+
+  uid() {
+    return this.path.uid();
+  }
+
+  isEqual = (other: SqValueNode): boolean => {
+    return this.uid() === other.uid();
+  };
+
+  sqValue(): SqValue | undefined {
+    return this.root.getSubvalueByPath(this.path, this.traverseCalculatorEdge);
+  }
+
+  parent() {
+    const parentPath = this.path.parent();
+    return parentPath
+      ? new SqValueNode(this.root, parentPath, this.traverseCalculatorEdge)
+      : undefined;
+  }
+
+  children() {
+    const value = this.sqValue();
+    if (!value) {
+      return [];
+    }
+    return getChildrenValues(value)
+      .map((childValue) => {
+        const path = childValue.context?.path;
+        return path
+          ? new SqValueNode(this.root, path, this.traverseCalculatorEdge)
+          : undefined;
+      })
+      .filter((a) => a !== undefined) as SqValueNode[];
```

Comment:
Slightly better, safer version:
```suggestion
      .filter((a): a is NonNullable<typeof a> => a !== undefined);
```

(it's annoying that TypeScript doesn't do it itself)